File: clients/mobile/assistant/lib/core/render_descriptor_v1.dart
// Dart projection that maps a spec 069 assistant_turn_v1 response into the
// canonical spec 073 render-descriptor-v1 shape (see
// `specs/073-web-mobile-assistant-frontend/design.md` § Render-Descriptor
// JSON Schema). Output ordering is canonical and identical to the JS
// reference at `web/pwa/lib/render_descriptor_v1.js`: body text, citations
// (source order), disambiguation choices (choice order), confirm
// accept/decline pair, retry action.
//
// This file is the shared core for both the iOS and Android adapter
// projections used by the spec 073 cross-language renderer canary
// (TP-073-03). It MUST NOT import any platform secure-storage, shared
// preferences, file-system, or path-provider API.

import 'generated/assistant_turn_v1.dart';

const String descriptorSchemaVersion = 'render-descriptor.v1';

/// Project a validated assistant turn response into a JSON-serializable
/// render-descriptor-v1 map. The input is validated via
/// `validateTurnResponse`; any unsupported response `schema_version`
/// throws before any node is emitted.
Map<String, dynamic> renderToDescriptorV1(Map<String, dynamic> response) {
  validateTurnResponse(response);

  final List<Map<String, dynamic>> nodes = <Map<String, dynamic>>[];

  final String body = response['body'] as String;
  if (body.isNotEmpty) {
    nodes.add(<String, dynamic>{'kind': 'text', 'text': body});
  }

  final List<dynamic> sources = response['sources'] as List<dynamic>;
  for (final dynamic raw in sources) {
    if (raw is! Map<String, dynamic>) continue;
    final Map<String, dynamic> node = <String, dynamic>{
      'kind': 'citation',
      'source_id': (raw['id'] ?? '').toString(),
      'label': (raw['title'] ?? '').toString(),
    };
    final dynamic url = raw['url'];
    if (url is String && url.isNotEmpty) {
      node['url'] = url;
    }
    nodes.add(node);
  }

  final dynamic dis = response['disambiguation_prompt'];
  if (dis is Map<String, dynamic>) {
    final String ref = (dis['disambiguation_ref'] ?? '').toString();
    final List<dynamic> choices = (dis['choices'] as List<dynamic>?) ?? const <dynamic>[];
    for (final dynamic c in choices) {
      if (c is! Map<String, dynamic>) continue;
      final dynamic number = c['number'];
      final int idx = number is int ? number : 0;
      nodes.add(<String, dynamic>{
        'kind': 'action',
        'action_kind': 'disambiguation_choice',
        'ref': ref,
        'label': (c['label'] ?? '').toString(),
        'choice_index': idx,
      });
    }
  }

  final dynamic confirm = response['confirm_card'];
  if (confirm is Map<String, dynamic>) {
    final String ref = (confirm['confirm_ref'] ?? '').toString();
    nodes.add(<String, dynamic>{
      'kind': 'action',
      'action_kind': 'confirm_accept',
      'ref': ref,
      'label': (confirm['positive_label'] ?? '').toString(),
    });
    nodes.add(<String, dynamic>{
      'kind': 'action',
      'action_kind': 'confirm_decline',
      'ref': ref,
      'label': (confirm['negative_label'] ?? '').toString(),
    });
  }

  final String errCause = response['error_cause'] as String;
  if (errCause.isNotEmpty) {
    nodes.add(<String, dynamic>{
      'kind': 'action',
      'action_kind': 'retry',
      'ref': errCause,
      'label': 'Retry',
    });
  }

  return <String, dynamic>{
    'schema_version': descriptorSchemaVersion,
    'nodes': nodes,
  };
}

File: web/pwa/lib/render_descriptor_v1.js
// Pure JS renderer that projects a spec 069 assistant_turn_v1 response
// into the spec 073 render-descriptor-v1 shape.
//
// The renderer is intentionally branch-only on response shape, never on
// scenario id, action class, platform, or transport_hint. Output ordering
// is canonical: body text, citations (source order), disambiguation choices
// (choice order), confirm accept/decline pair, retry action.
//
// This module is the canonical JS reference consumed by:
//   - the planned web /assistant client renderer (spec 073 SCOPE-2)
//   - the cross-language renderer canary at
//     tests/unit/clients/render_descriptor_canary_test.go (TP-073-03)
'use strict';

const DESCRIPTOR_SCHEMA_VERSION = 'render-descriptor.v1';
const RESPONSE_SCHEMA_VERSION = 'v1';

function renderToDescriptorV1(response) {
  if (response === null || typeof response !== 'object' || Array.isArray(response)) {
    throw new Error('render_descriptor_v1: response must be an object');
  }
  if (response.schema_version !== RESPONSE_SCHEMA_VERSION) {
    throw new Error(
      'render_descriptor_v1: unsupported response schema_version: ' +
        String(response.schema_version)
    );
  }
  const nodes = [];

  const body = typeof response.body === 'string' ? response.body : '';
  if (body.length > 0) {
    nodes.push({ kind: 'text', text: body });
  }

  const sources = Array.isArray(response.sources) ? response.sources : [];
  for (const s of sources) {
    if (s === null || typeof s !== 'object') continue;
    const node = {
      kind: 'citation',
      source_id: String(s.id == null ? '' : s.id),
      label: String(s.title == null ? '' : s.title),
    };
    if (typeof s.url === 'string' && s.url.length > 0) {
      node.url = s.url;
    }
    nodes.push(node);
  }

  const dis = response.disambiguation_prompt;
  if (dis !== null && typeof dis === 'object' && !Array.isArray(dis)) {
    const ref = String(dis.disambiguation_ref == null ? '' : dis.disambiguation_ref);
    const choices = Array.isArray(dis.choices) ? dis.choices : [];
    for (const c of choices) {
      if (c === null || typeof c !== 'object') continue;
      const idx = typeof c.number === 'number' ? c.number : 0;
      nodes.push({
        kind: 'action',
        action_kind: 'disambiguation_choice',
        ref: ref,
        label: String(c.label == null ? '' : c.label),
        choice_index: idx,
      });
    }
  }

  const conf = response.confirm_card;
  if (conf !== null && typeof conf === 'object' && !Array.isArray(conf)) {
    const ref = String(conf.confirm_ref == null ? '' : conf.confirm_ref);
    nodes.push({
      kind: 'action',
      action_kind: 'confirm_accept',
      ref: ref,
      label: String(conf.positive_label == null ? '' : conf.positive_label),
    });
    nodes.push({
      kind: 'action',
      action_kind: 'confirm_decline',
      ref: ref,
      label: String(conf.negative_label == null ? '' : conf.negative_label),
    });
  }

  const err = typeof response.error_cause === 'string' ? response.error_cause : '';
  if (err.length > 0) {
    nodes.push({
      kind: 'action',
      action_kind: 'retry',
      ref: err,
      label: 'Retry',
    });
  }

  // Spec 075 SCOPE-075-06.3 — optional legacy-retirement notice rendered
  // as a one-line text addendum AFTER the primary body so it never
  // blocks the assistant response. Field is additive and optional on
  // the v1 wire contract; absent on every turn that did not match a
  // retired command.
  const notice = response.notice;
  if (notice !== null && typeof notice === 'object' && !Array.isArray(notice)) {
    const replacement = typeof notice.replacement_example === 'string' ? notice.replacement_example : '';
    if (replacement.length > 0) {
      nodes.push({ kind: 'text', text: replacement });
    }
  }

  return {
    schema_version: DESCRIPTOR_SCHEMA_VERSION,
    nodes: nodes,
  };
}

module.exports = { renderToDescriptorV1, DESCRIPTOR_SCHEMA_VERSION };
